### ISSUE TITLE
Problem with samtools

### ISSUE THREAD

**Issue by rtmag** [mhammell-laboratory/tetoolkit#11]
Hello!

there seems to be a problem with samtools, any idea on what might be the problem?

Cheers!

```
INFO  @ Thu, 09 Mar 2017 03:03:27: Done building gene index ......

INFO  @ Thu, 09 Mar 2017 03:03:33:
Building TE index .......

INFO  @ Thu, 09 Mar 2017 03:13:14: Done building TE index ......

INFO  @ Thu, 09 Mar 2017 03:13:14:
Reading sample files ...

Error occured when reading first line of sample file /home/rtm/SJlab/deepa/bam/HCT_siK_Aligned.sortedByCoord.out.bam.
Error: 'samtools returned with error 1: stdout=, stderr=[bam_sort] Use -T PREFIX / -o FILE to specify temporary and final output files\nUsage: samtools sort [options...] [in.bam]\nOptions:\n  -l INT     Set compression level, from 0 (uncompressed) to 9 (best)\n  -m INT     Set maximum memory per thread; suffix K/M/G recognized [768M]\n  -n         Sort by read name\n  -o FILE    Write final output to FILE rather than standard output\n  -T PREFIX  Write temporary files to PREFIX.nnnn.bam\n  -@, --threads INT\n             Set number of sorting and compression threads [1]\n      --input-fmt-option OPT[=VAL]\n               Specify a single input file format option in the form\n               of OPTION or OPTION=VALUE\n  -O, --output-fmt FORMAT[,OPT[=VAL]]...\n               Specify output format (SAM, BAM, CRAM)\n      --output-fmt-option OPT[=VAL]\n               Specify a single output file format option in the form\n               of OPTION or OPTION=VALUE\n      --reference FILE\n               Reference sequence FASTA FILE [null]\n'
[Exception type: SamtoolsError, raised in utils.py:75]
```

**Comment by olivertam** [comment]
Hi. This looks like an issue with the updated pysam (>0.9) and samtools (>1.3) where the old `samtools sort` command no longer works. We are in the process of adding the ability to handle a newer pysam and samtools, and should hopefully get this into the code soon.
Thanks, and apologies for the slow responses.

**Comment by retrogenomics** [comment]
Hi,
I've got the same problem. I was wondering if you had found a work around since then?
Thank you in advance

**Comment by yingjin07** [comment]
Hi, there is a work around using the samtools to sort the bam files according to read name before running TEtranscripts. Thanks.

**Comment by retrogenomics** [comment]
It worked. Thanks

**Comment by vasilislenis** [comment]
@retrogenomics 
What do you mean it worked? 
I'm getting the same error. Could you please tell me how you fixed it?
Thank you very much in advance.

**Comment by retrogenomics** [comment]
@vasilislenis It worked as suggested by @yingjin07, i.e. by using samtools to sort the bam files according to read names rather than by coordinates (`samtools sort -@4 -O BAM -n file.bam -o file.sortedByReadname.bam`) . Then you can run TEtranscripts without getting the error.

**Comment by vasilislenis** [comment]
@retrogenomics Thank you very much for your reply but unfortunately, it didn't work. I'm still getting the same error.

**Comment by olivertam** [comment]
@vasilislenis There should be only one count table (both coding genes and TE in one). You can easily separate them out by searching for the `:` in the name (which should only appear in TE). It is not immediately clear what the error is (other than a missing file). Are you using the test data, or your own? If you don't mind providing the command line (you can replace any file names with placeholders if you prefer), that would be great. Thanks.

**Comment by olivertam** [comment]
@vasilislenis Also, was a `[prefix]_DESeq.R` file generated in addition to the counts table?

**Comment by vasilislenis** [comment]
@olivertam. I am really sorry! I forgot to load the R module.
Based on the log file everything went well. 
It was the test data, so now I will try with mine. I'll let you know how it goes.
Thank you very much!

**Comment by wjyzidane** [comment]
I have the same problem:
[Exception type: SamtoolsError, raised in utils.py:75]

so I sort them by name and re-run TEtranscript, but got another error:

If the BAM file is sorted by coordinates, please specify --sortByPos and re-run!

**Comment by olivertam** [comment]
Hi.
If you could paste a copy of the log file for the second run, and perhaps the header and the first 10 or so alignments of the sorted BAM file, I can take a closer look.
Thanks.

**Comment by wjyzidane** [comment]
Thanks!

Here is my file:

![image](https://user-images.githubusercontent.com/34944811/38147343-db7f4626-341f-11e8-900c-3b3a89885446.png)

Here is the logfile:
![image](https://user-images.githubusercontent.com/34944811/38147662-40ca6f46-3421-11e8-9601-26c252bf2f28.png)

Jingyi

**Comment by olivertam** [comment]
Hi Jingyi,
Looking at the BAM file, it is odd that it still looks like it is sorted by co-ordinates. I see chr1 for the first batch of alignments, and they appear to be arranged in numeric order (see column 2). You can also see that although the first 8 lines appear to be pairs of alignments, line 9 and line 12 appear to be a pair, but are separated by two other alignments.
It might be worth double-checking the headers to see if the files are sorted correctly. The header should have a line that looks like the following: 
`@HD    VN1.4     SO:queryname`
and not
`@HD    VN1.4     SO:coordinate`
Thanks.

**Comment by wjyzidane** [comment]
Thanks so much. It is indeed sorted by coordinate. Let me re-sort them again.

Have a good weekend.
Jingyi

**Comment by olivertam** [comment]
Issue relating to the newer version of samtools/pysam should be addressed in TEToolkit v2.0.1. Note that pysam requirement is now v0.9.0 or higher
